<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [545]:
import pandas as pd
import psycopg2

In [547]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [548]:
query_3_1 = f'''
SELECT *
FROM vacancies
LIMIT {10}
'''

In [549]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   
5  55906072                                          Корректор   
6  55523207                   React Front-end Middle Developer   
7  55050261                                Инженер-программист   
8  55583432                            Системный администратор   
9  55312414        Ведущий специалист группы разработки систем   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   
5  Корректура текстов\tГрамотность\tРабота в кома...       Полный день   
6       JavaScript\tTypeScript\tCSS3\tNode.js\tHTML5     Гибкий график   
7  Java SE\tSpring Framework\tSpring Boot\tGit\tS...       Полный день   
8  Linux\tАдминистрирование сетевого оборудования...       Полный день   
9  1С программирование\t1С: Предприятие 8\t1С: До...       Полный день   

           experience           employment  salary_from  salary_to  area_id  \
0           Нет опыта     Полная занятость      64000.0        NaN     1516   
1       От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2       От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3       От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4           Нет опыта  Частичная занятость      15000.0        NaN     1955   
5           Нет опыта     Полная занятость          NaN        NaN     2323   
6  От 1 года до 3 лет     Полная занятость          NaN        NaN     2759   
7           Нет опыта     Полная занятость      40000.0    60000.0       49   
8  От 1 года до 3 лет     Полная занятость      50000.0        NaN        3   
9       От 3 до 6 лет     Полная занятость          NaN        NaN       47   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740  
5       584934  
6       204511  
7      4448636  
8        55126  
9      1413754

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [550]:
query_3_2 = f'''
SELECT*
FROM employers
LIMIT {10}
'''

In [551]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


id                                               name  area
0     2393                   Программный Продукт, ИТ-компания     1
1    72977                                          БАРС Груп    88
2     3155  Бест, Торгово-производственная компания, Екате...     3
3      675                                   КОРУС Консалтинг     2
4  1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88
5   213397                                            МойОфис     1
6    41862                                             Контур     3
7  1080425  Челябинский государственный университет, ФГБОУ...   104
8     4080                           Axenix (ранее Accenture)     1
9     3415                                       ИНФОРМЗАЩИТА     1

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [552]:
query_3_3 = f'''
SELECT *
FROM areas
LIMIT {10}
'''

In [553]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


id                           name
0  2758                        Тбилиси
1     8                         Майкоп
2  1180                       Нерюнгри
3  1240                    Новокузнецк
4     2                Санкт-Петербург
5     1                         Москва
6  2061  Подольск (Московская область)
7  1545                   Новочеркасск
8  2759                        Ташкент
9  3695                   Промышленная

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [554]:
query_3_4 = f'''
SELECT *
FROM industries
LIMIT {10}
'''

In [555]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


id                                               name
0   7.540                Разработка программного обеспечения
1   7.539  Системная интеграция,  автоматизации технологи...
2  27.550              Безалкогольные напитки (производство)
3  27.551  Безалкогольные напитки (продвижение, оптовая т...
4  13.664             Управление и эксплуатация недвижимости
5   5.466                                   Складские услуги
6  39.442                             Вуз, ссуз колледж, ПТУ
7  44.393                              Консалтинговые услуги
8   5.467                      Курьерская, почтовая доставка
9  41.525                                   Интернет-магазин

***

### Выводы по предварительному анализу данных:
* База данных содержит достаточно большое количество вакансий (49197). Это может указывать на активность рынка труда и наличие разнообразных возможностей для соискателей.
* Количество работодателей в базе составляет 23501. Это означает, что на рынке труда представлено большое количество компаний и организаций, которые предлагают вакансии.
* В базе данных содержится информация о 1362 регионах. Это указывает на географическое разнообразие вакансий и возможности трудоустройства в различных местах.
* Сферы деятельности представлены в базе данных 294. Это свидетельствует о разнообразии отраслей и секторов экономики, в которых существуют вакансии.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [556]:
query_4_1 = f'''
SELECT 
    COUNT(*) as cnt,
    a.name
FROM vacancies as v
JOIN areas as a on v.area_id=a.id
GROUP BY 2
ORDER BY 1 desc
LIMIT 5
'''

In [557]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             name
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [558]:
query_4_2 = f'''
SELECT
    COUNT(*) as Количество
FROM vacancies as v
WHERE v.salary_from is not null or v.salary_to is not null
'''

In [559]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество
0       24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [560]:
query_4_3 = f'''
SELECT
    round(avg(salary_from), 0) as "Нижняя граница",
    round(avg(salary_to), 0) as "Верхняя граница"
FROM vacancies as v

'''

In [561]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Нижняя граница  Верхняя граница
0         71065.0         110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [562]:
query_4_4 = f'''
SELECT
    COUNT(*),
    schedule,
    employment
FROM vacancies as v
GROUP BY 2, 3
ORDER BY 1 desc
'''

In [563]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [564]:
query_4_5 = f'''
    SELECT 
        COUNT(v.id) AS cnt,
        v.experience AS "Требуемый опыт работы"
    FROM vacancies v
    GROUP BY v.experience
    ORDER BY 1
    '''

In [565]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


cnt Требуемый опыт работы
0   1337           Более 6 лет
1   7197             Нет опыта
2  14511         От 3 до 6 лет
3  26152    От 1 года до 3 лет

***

### Выводы по детальному анализу вакансий:
* Пятёрка лидеров по количеству вакансий, включает Москву, Санкт-Петербург, Минск, Новосибирск и Алматы. Это города, в которых предложение вакансий наиболее высоко, что может указывать на их значимость и развитость как центров бизнеса и экономики.
* Информация о зарплате указана для 24073 вакансий. Это позволяет провести анализ зарплатного уровня и оценить рыночную стоимость работы.
* Среднее значение для нижней зарплатной вилки составляет 71065, а для верхней зарплатной вилки - 110537. Это диапазон зарплат, предлагаемых во многих вакансиях, и может служить ориентиром для соискателей в планировании своей заработной платы.
* Опыт работы от 1 года до 3 лет является наиболее востребованным. Это может указывать на то, что многие работодатели ищут кандидатов с определенным уровнем опыта, который считается оптимальным для выполнения задач в соответствующей области.
* Полная занятость является наиболее популярным типом занятости. Это означает, что большинство работодателей ищут сотрудников, готовых работать на полный рабочий день.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [566]:
query_5_1 = f'''

(SELECT
    e.name,
    COUNT(v.name)
FROM employers as e
JOIN vacancies as v on v.employer_id=e.id
GROUP BY 1
ORDER BY 2 desc
LIMIT 1)
UNION
(SELECT
    e.name,
    COUNT(v.name)
FROM employers as e
JOIN vacancies as v on v.employer_id=e.id
GROUP BY 1
ORDER BY 2 desc
OFFSET 4
LIMIT 1)
ORDER BY 2 desc
'''

In [567]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [568]:
query_5_2 = f'''

SELECT DISTINCT
    a.name, 
    COUNT(e.id) as "Количество работодателей", 
    COUNT(v.id) as "Количество вакансий"
FROM areas as a
LEFT JOIN vacancies as v on a.id=v.area_id
LEFT JOIN employers e on e.area = a.id
GROUP BY 1
HAVING COUNT(v.id) = 0
ORDER BY 2 desc
LIMIT 5
'''

In [569]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  Количество работодателей  Количество вакансий
0              Россия                       410                    0
1           Казахстан                       207                    0
2  Московская область                        75                    0
3  Краснодарский край                        19                    0
4            Беларусь                        18                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [570]:
query_5_3 = f'''
SELECT 
    e.name as "Работодатель", 
    COUNT(DISTINCT v.area_id) as "Количество регионов"
FROM vacancies as v
LEFT JOIN areas as a on a.id=v.area_id
LEFT JOIN employers e on v.employer_id = e.id
GROUP BY 1
ORDER BY 2 desc
LIMIT 5
'''

In [571]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Работодатель  Количество регионов
0                  Яндекс                  181
1              Ростелеком                  152
2              Спецремонт                  116
3  Поляков Денис Иванович                   88
4               ООО ЕФИН                    71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [572]:
query_5_4 = f'''
SELECT 
    COUNT(e.name)
FROM employers as e 
LEFT JOIN employers_industries AS ei ON e.id=ei.employer_id
WHERE
    ei.employer_id is NULL
'''

In [573]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [574]:
query_5_5 = f'''
    SELECT
        e.name AS "Названия компаний",
        COUNT(ei.industry_id) AS "Кол-во сфер деятельности"
    FROM employers e
    FULL JOIN employers_industries ei ON ei.employer_id = e.id
    GROUP BY e.name
    HAVING COUNT(ei.industry_id) = 4
    ORDER BY 1
    OFFSET 2
    LIMIT 1
    '''

In [575]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Названия компаний  Кол-во сфер деятельности
0              2ГИС                         4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [576]:
query_5_6 = f'''
    SELECT
        COUNT(ei.employer_id) AS "Кол-во работодателей"
    FROM employers_industries ei
    LEFT JOIN industries i ON ei.industry_id = i.id
    WHERE i.name = 'Разработка программного обеспечения'
    '''

In [577]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Кол-во работодателей
0                  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [578]:
query_5_7 = f'''
    SELECT
        a.name AS "Регионы-миллионники",
        COUNT(v.id) "Кол-во вакансий"
    FROM vacancies v
    LEFT JOIN employers e on v.employer_id = e.id
    LEFT JOIN areas a on v.area_id = a.id
    WHERE e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 
    'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
    GROUP BY a.name

    UNION ALL

    SELECT 
        'Total', 
        COUNT(*)
    FROM vacancies v
    LEFT JOIN employers e on v.employer_id = e.id
    LEFT JOIN areas a on v.area_id = a.id
    WHERE e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 
    'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
    ORDER BY 2
    '''

In [579]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Регионы-миллионники  Кол-во вакансий
0                 Омск               21
1            Челябинск               22
2           Красноярск               23
3            Волгоград               24
4                Пермь               25
5               Казань               25
6       Ростов-на-Дону               25
7                  Уфа               26
8               Самара               26
9            Краснодар               30
10             Воронеж               32
11         Новосибирск               35
12     Нижний Новгород               36
13        Екатеринбург               39
14     Санкт-Петербург               42
15              Москва               54
16               Total              485

***


## Выводы по анализу работодателей:
* Компания Яндекс занимает первое место по количеству вакансий. Это указывает на значительное присутствие и активность компании на рынке труда.
* В регионах, где нет вакансий, большинство работодателей находится в России. Это может быть связано с ориентацией этих работодателей на местный рынок труда.
* Выявлено 8419 работодателей, у которых не указана сфера деятельности. Это может указывать на неполную или некорректную заполненность данных в базе.
* Наиболее востребованный опыт работы - от 1 года до 3 лет. Это может быть связано с тем, что работодатели ищут кандидатов, которые имеют определенный опыт работы, но еще не являются слишком специализированными.
* Полная занятость является наиболее популярным типом занятости. Это указывает на то, что работодатели предпочитают нанимать сотрудников на полный рабочий день.
* Компания Яндекс представила 485 вакансий в 17 городах-миллионниках. Это говорит о широком географическом охвате и наличии возможностей для трудоустройства в разных городах.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [580]:
query_6_1 = f'''
    SELECT
        COUNT(v.id) AS "Кол-во вакансий, связанных с DS"
    FROM vacancies v
    WHERE lower(v.name) LIKE '%данн%' OR lower(v.name) LIKE '%data%'
    '''

In [581]:
df =pd.read_sql_query(query_6_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3746746274.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df =pd.read_sql_query(query_6_1, connection)


Кол-во вакансий, связанных с DS
0                             1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [582]:
query_6_2 = f'''
    SELECT
        COUNT(v.id) AS "Кол-во вакансий junior DS"
    FROM vacancies v
    WHERE (lower(v.name) LIKE '%data scientist%'
    OR lower(v.name) LIKE '%data science%'
    OR lower(v.name) like '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%')
    OR lower(v.name) LIKE '%machine learning%'
    OR lower(v.name) LIKE '%машинн%обучен%')
    AND 
    (lower(v.name) LIKE '%junior%'
    OR lower(v.experience) LIKE '%нет опыта%'
    OR lower(v.experience) LIKE '%стажировка%')
    '''

In [583]:
df =pd.read_sql_query(query_6_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\570157465.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df =pd.read_sql_query(query_6_2, connection)


Кол-во вакансий junior DS
0                         51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [584]:
query_6_3 = f'''
SELECT 
     COUNT(v.id)
FROM vacancies v
WHERE ((v.key_skills like '%SQL%') or (v.key_skills ilike '%postgres%'))

    AND ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%') 
    OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL))
'''   

In [585]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [586]:
query_6_4 = f'''
SELECT
        COUNT(*)
FROM vacancies v
WHERE
        (
        name ilike '%data scientist%'  OR
        name ilike  '%data science%' OR
        name ilike  '%исследователь данных%'  OR
        (name ilike  '%ML%' and name not ilike '%HTML%')  OR
        name ilike '%machine learning%' OR  
        name ilike '%машинн%обучен%'
        )
        AND
        key_skills ilike '%Python%'
                
'''   

In [587]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [588]:
query_6_5 = f'''
SELECT
      round(avg(LENGTH(v.key_skills) - 
      LENGTH(replace(v.key_skills, chr(9), '')) + 1), 2) as "Среднее к-во скилов"
          
FROM vacancies v
WHERE (lower(v.name) like '%data scientist%' OR 
      lower(v.name) like '%data science%' OR
      lower(v.name) like '%исследователь данных%' OR
      (v.name like '%ML%' and v.name not like '%HTML%') OR
      lower(v.name) like '%machine learning%' OR
      lower(v.name) like '%машинн%обучен%%')                 
'''   

In [589]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее к-во скилов
0                 6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [590]:
query_6_6 = f'''
SELECT
      round(avg(coalesce((salary_from+salary_to)/2, 
      salary_from, salary_to))) "Средняя ЗП DS с опытом 3-6 лет"
    
FROM vacancies v
WHERE (lower(v.name) like '%data scientist%' OR 
      lower(v.name) like '%data science%' OR
      lower(v.name) like '%исследователь данных%' OR
      (v.name like '%ML%' and v.name not like '%HTML%') OR
      lower(v.name) like '%machine learning%' OR
      lower(v.name) like '%машинн%обучен%%') 
      AND (v.salary_from is not null or salary_to is not null)   
      AND (v.experience = 'От 3 до 6 лет')                       
'''   

In [591]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Средняя ЗП DS с опытом 3-6 лет
0                        243115.0

***

### Выводы по предметному анализу:
* Количество вакансий, связанных с работой с данными, составляет 1771. Это указывает на наличие значительного спроса на специалистов в области анализа данных и data science.
* Существует 51 подходящая вакансия для начинающего дата-сайентиста. Это означает, что существуют возможности для новичков в этой области найти работу и начать свою карьеру.
* Количество вакансий для дата-сайентистов, где SQL или postgre указаны в качестве ключевых навыков, составляет 201. Это указывает на важность знания SQL или postgre для работы в этой области и требования работодателей в этом отношении.
* Количество вакансий, где Python указывается в качестве ключевого навыка, составляет 351. Это говорит о том, что знание Python востребовано в работе дата-сайентиста и является одним из основных инструментов в этой области.
* Среднее количество ключевых навыков для дата-сайентиста составляет более 6. Это говорит о том, что в этой области требуются специалисты с широким спектром навыков и компетенций.
* Средняя зарплата для дата-сайентиста с опытом работы от 3 до 6 лет составляет 243115. Это указывает на то, что опытные специалисты в этой области могут рассчитывать на высокую заработную плату.

***

### Дополнительные исследования

7.1 Проведём анализ количества вакансий по сферам деятельности. Выведем ТОП-10 сфер деятельности по количеству вакансий.

In [592]:
query_7_1 = f'''
SELECT
    i.name AS "Сфера деятельности",
    COUNT(v.id) AS "Количество вакансий"
FROM industries i
LEFT JOIN employers_industries ei ON i.id = ei.industry_id
LEFT JOIN vacancies v ON ei.employer_id = v.employer_id
GROUP BY i.name
ORDER BY COUNT(v.id) DESC
LIMIT 10
'''

In [593]:
df =pd.read_sql_query(query_7_1, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\1669388331.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df =pd.read_sql_query(query_7_1, connection)


Сфера деятельности  Количество вакансий
0                Разработка программного обеспечения                12499
1  Системная интеграция,  автоматизации технологи...                11034
2  Интернет-компания (поисковики, платежные систе...                 6413
3                                               Банк                 2742
4                                    Мобильная связь                 1585
5                                 Кадровые агентства                 1143
6                              Консалтинговые услуги                 1137
7                                 Интернет-провайдер                 1095
8                       Розничная сеть (продуктовая)                 1034
9                                Фиксированная связь                 1028

7.2 Проведём анализ количества работодателей по сферам деятельности. Выведем ТОП-10 сфер деятельности по количеству работодателей.

In [594]:
query_7_2 = f'''
SELECT
    i.name AS "Сфера деятельности",
    COUNT(DISTINCT ei.employer_id) AS "Количество работодателей"
FROM employers_industries ei
JOIN industries i ON ei.industry_id = i.id
GROUP BY i.name
ORDER BY COUNT(DISTINCT ei.employer_id) DESC
LIMIT 10
'''

In [595]:
df =pd.read_sql_query(query_7_2, connection)
df

C:\Users\user\AppData\Local\Temp\ipykernel_11672\3680141440.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df =pd.read_sql_query(query_7_2, connection)


Сфера деятельности  Количество работодателей
0                Разработка программного обеспечения                      3553
1  Системная интеграция,  автоматизации технологи...                      2993
2  Интернет-компания (поисковики, платежные систе...                      1675
3  Маркетинговые, рекламные, BTL, дизайнерские, E...                       798
4                              Консалтинговые услуги                       662
5                                 Кадровые агентства                       389
6  Электронно-вычислительная, оптическая, контрол...                       374
7                       Розничная сеть (продуктовая)                       368
8                                   Интернет-магазин                       351
9                        Архитектура, проектирование                       346

In [596]:
#закрываем соединение после окончания работы
connection.close()

### Выводы по дополнительным исследованиям:
* Сфера деятельности "Разработка программного обеспечения" имеет наибольшее количество вакансий (12499) и наибольшее количество работодателей (3553). Это указывает на высокий спрос на специалистов в этой области и наличие множества компаний, занимающихся разработкой программного обеспечения.
* Сфера деятельности "Системная интеграция, автоматизации технологий" также имеет значительное количество вакансий (11034) и работодателей (2993). Это говорит о том, что автоматизация технологических процессов и системная интеграция являются востребованными и перспективными областями.
* Сфера деятельности "Интернет-компания (поисковики, платежные системы и другие)" имеет меньшее количество вакансий (6413) и работодателей (1675) по сравнению с предыдущими двумя сферами. Однако она все равно представляет собой значительную долю рынка труда и может предоставлять интересные возможности для трудоустройства.

***

# Общий вывод по проекту

### Общий вывод по проекту:
* Выборка содержит 49197 вакансий, представлен 23501 работодатель, содержится информация о 1362 регионах по 294 сферам деятельности.
* Лидерами по количеству вакансий являются мегаполисы, выделяя их значимость на рынке труда и развитость как центров бизнеса и экономики.
* Среднее значение для нижней зарплатной вилки составляет 71065, а для верхней зарплатной вилки - 110537.
* Опыт работы от 1 года до 3 лет является наиболее востребованным.
* Полная занятость является наиболее популярным типом занятости.
* Компания "Яндекс" занимает первое место по количеству вакансий - 1933.
* Полная занятость является наиболее популярным типом занятости.
* Количество вакансий, связанных с работой с данными, составляет 1771. Количество вакансий, где Python указывается в качестве ключевого навыка, составляет 351. Среднее количество ключевых навыков для дата-сайентиста составляет более 6, при этом соискатель может расчитывать в среднем на зарплату в 243115 руб.
* Сфера деятельности "Разработка программного обеспечения" имеет наибольшее количество вакансий - 12499 и наибольшее количество работодателей - 3553.